atom/renderer/api/atom_api_renderer_ipc.cc

```diff
@@ -62,6 +66,20 @@ class IPCRenderer : public mate::Wrappable<IPCRenderer> {
     electron_browser_ptr_->Message(internal, channel, arguments.Clone());
   }
 
+  v8::Local<v8::Promise> Invoke(mate::Arguments* args,
+                                const std::string& channel,
+                                const base::Value& arguments) {
+    atom::util::Promise p(args->isolate());
+    auto handle = p.GetHandle();
+    electron_browser_ptr_->Invoke(
```

Comment:
Although probs not as clean as this, IMO we should proxy thrown errors over this as well and reject the promise

Reply:
Hmmmm maybe, but it's totally possible (likely even) that errors won't happen synchronously in the handler. e.g.

```
ipcMain.on('read-some-file', (e) => {
  fs.readFile('some-file', (err, data) => e.reply(data))
})
```

there's no way we can automatically catch that kind of error. But i guess there should be a way for the caller to send a rejection? and we can at least catch synchronous errors if they happen.

Reply:
@nornagon That's why I was suggesting a `setHandler` method that provided a channel and an async function.  Then we should be able to catch async errors as well as it's just a promise.

```
ipcMain.setHandler = (channel, fn) => {
  if (ipcMain.handlers(channel).length !== 0) throw err;
  ipcMain.on(channel, async (...args) => {
    try {
      event.reply(await Promise.resolve(fn(...args)))
    } catch (err) {
      event.sendError(err);
    }
  });
}

ipcMain.removeHandler = (channel, fn) => {
  // ...
}
```

Reply:
I mean, assuming that the caller uses promises :)

Reply:
@nornagon That's why I wrapped it in `Promise.resolve(fn(...args))`, whether it returns promises or not that will work 👍

Reply:
You have to use promises all the way down for exceptions to be correctly propagated, though. All I'm saying is that the handler can never be 100% guaranteed to call `reply` or `sendError` because it's under the control of the user.

Reply:
I mean if the user isn't handling errors / leaving dangling / un-awaited promises that's on them.  We can't catch everything, but we can make it really easy to catch everything if you write a proper async handler

Reply:
If we build a `setHandler` API like you propose, do you think it would be worth separating the channel surface so it wasn't just handled with `ipcMain.on` but rather through a separate event bus?

Reply:
> but rather through a separate event bus?

I think so yes

Reply:
On further thought, this approach means that it's not possible to write a handler in a promise-less asynchronous style, like

```js
ipcMain.handle('thing', (e, arg) => {
  doAsyncThing(arg, (result) => e.reply(result))
})
```

because the return value of the handler, if not a promise, will be immediately sent as a response. You'd have to write

```js
ipcMain.handle('thing', (e, arg) => {
  return new Promise(resolve => {
    doAsyncThing(arg, resolve)
  })
})
```

or

```js
ipcMain.handle('thing', async (e, arg) => {
  await new Promise(resolve => doAsyncThing(arg, resolve))
})
```

or some similar incantation.

I like promises and think we should support them, but I'm a little worried about adding a core API that doesn't support not-Promises.

Reply:
@nornagon I think it's ok to be promise-only, that's what all our other APIs are moving towards (removing callbacks, using promises).

If folks really want to use promises they can do `new Promise((resolve)=>{})` as you outlined above or they can just promisify their handler.  E.g.

```js
const { promisify } = require('util');

ipcMain.handle('thing', promisify((cb) => {
  doAsyncWork(arg, cb);
});
```

Reply:
Ok, I did it :) PTAL